frontend/packages/topology/src/components/page/DroppableTopologyComponent.tsx

```diff
@@ -0,0 +1,62 @@
+import * as React from 'react';
+import { NativeTypes } from 'react-dnd-html5-backend';
+import { DropTarget } from 'react-dnd';
+import { DropTargetMonitor } from 'react-dnd/lib/interfaces';
+import { Model } from '@patternfly/react-topology';
+import { history } from '@console/internal/components/utils';
+import withDragDropContext from '@console/internal/components/utils/drag-drop-context';
+import {
+  JarFileContextType,
+  JarFileContext,
+} from '@console/internal/components/jar-file/jar-file-context';
+import { ConnectedTopologyView } from './TopologyView';
+import { TopologyViewType } from '../../topology-types';
+import { isJARFile } from '../../utils';
+
+const boxTarget = {
+  drop(props, monitor) {
+    if (props.onDrop && monitor.isOver()) {
+      props.onDrop(props, monitor);
+    }
+  },
+};
+
+const DroppableTopology = DropTarget(NativeTypes.FILE, boxTarget, (connectObj, monitor) => {
+  return {
+    connectDropTarget: connectObj.dropTarget(),
+    isOver: monitor.isOver(),
+    canDrop: monitor.canDrop(),
+  };
+})(ConnectedTopologyView);
+
+export const DroppableTopologyComponent = withDragDropContext<DroppableTopologyComponentProps>(
+  (props) => {
+    const { setJarFileObj } = React.useContext<JarFileContextType>(JarFileContext);
+
+    const handleFileDrop = (item, monitor: DropTargetMonitor) => {
+      if (!monitor) {
+        return;
+      }
+      const [file] = monitor.getItem().files;
+      if (!file) {
+        return;
+      }
+
+      if (!isJARFile(file)) {
+        return;
+      }
+      if (isJARFile(file)) {
+        setJarFileObj(file);
+        history.push(`/upload-jar/ns/${props.namespace}`);
+      }
+    };
```

Comment:
We want to support helm chart archives in the future and therefore need to create an extension for different file types. This is also needed because topology is becoming a base component which dev-console extends. Jar file upload is considered a dev-console component and needs to be decoupled from topology.

Since the JAR upload extension needs access to a context, we either create a generic context which simply holds any file, or we need to have the extension define a custom hook. Needing to render a hook complicates things since we're execute the handler in a side effect. I think we can turn the context you already created into a generic context and move the logic for reading the extensions in it.

We could simply create an extension as follows:
```
type FileUploadExtension = {
  // array of file extensions supported
  fileExtensions: string[],
  // function which returns `true` if the file is supported, `false` otherwise
  // handler may choose to redirect to a new URL if supported
  handler: (file: File, namespace: String) => boolean;
}
```

```
fileExtension: ['jar'],
handler: (file: File, namespace: string) => {
  history.push(`/upload-jar/ns/${namespace}`)
  return true;
},
```

Topology can simply call the `setFileUpload` function on the context and the context will loop over the extensions. If a handler is found, it will set the file in the context, if not it can raise the toast.

This would make topology simply a gateway to a generic file upload facility which can be invoked by anyone else.

The context should probably also provide info to consumers such as whether or not there are any file upload extensions registered. If not, the dropzone should be disabled.

Reply:
Thank you for explaining all this. I have added an extension for file upload. PTAL